h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -0,0 +1,172 @@
+        stf.leftChildren[0] = sharedTreeSubgraph.rootNode.getLeftChild() != null ? sharedTreeSubgraph.rootNode.getLeftChild().getNodeNumber() : -1;
+
+        List<SharedTreeNode> nodesToTraverse = new ArrayList<>();
+        nodesToTraverse.add(sharedTreeSubgraph.rootNode);
+        append(stf.rightChildren, stf.leftChildren, stf.thresholds, nodesToTraverse, -1);
+
+        return stf;
+    }
+
+    /**
+     * Creates a map of {@link SharedTreeNode} where number of the node is the key.
+     *
+     * @param sharedTreeNodes A {@link List} of {@link SharedTreeNode} to create map from
+     * @return An non-thread-safe instance of {@link Map} with node number being the key to a {@link SharedTreeNode}
+     */
+    private static Map<Integer, SharedTreeNode> createTreeNodeMap(List<SharedTreeNode> sharedTreeNodes) {
+        Map<Integer, SharedTreeNode> treeNodeMap = new HashMap<>(sharedTreeNodes.size());
+
+        for (SharedTreeNode sharedTreeNode : sharedTreeNodes) {
+            treeNodeMap.put(sharedTreeNode.getNodeNumber(), sharedTreeNode);
+        }
+
+        assert treeNodeMap.size() == sharedTreeNodes.size(); // Verify there are no duplicated tree nodes (nodes with the same node number)
+
+        return treeNodeMap;
+    }
+
+    private static void append(final int[] rightChildren, final int[] leftChildren, final float[] thresholds,
+                               List<SharedTreeNode> nodesToTraverse, int pointer) {
+        if(nodesToTraverse.isEmpty()) return;
+
+        List<SharedTreeNode> discoveredNodes = new ArrayList<>();
+
+        for (SharedTreeNode node : nodesToTraverse) {
+            pointer++;
+            final SharedTreeNode leftChild = node.getLeftChild();
+            final SharedTreeNode rightChild = node.getRightChild();
+            thresholds[pointer] = node.getSplitValue();
+            final String[] inclusiveLevels = getInclusiveLevels(node);// How to compress 'em & send them to FE ? ?
+            if (leftChild != null) {
+                discoveredNodes.add(leftChild);
+                leftChildren[pointer] = leftChild.getNodeNumber();
+            } else {
+                leftChildren[pointer] = NO_CHILD;
+            }
+
+            if (rightChild != null) {
+                discoveredNodes.add(rightChild);
+                rightChildren[pointer] = rightChild.getNodeNumber();
+            } else {
+                rightChildren[pointer] = NO_CHILD;
+            }
+        }
+
+        append(rightChildren, leftChildren, thresholds, discoveredNodes, pointer);
+    }
+
+    private static String[] getInclusiveLevels(final SharedTreeNode node) {
+        if (!node.isBitset() || node.getInclusiveLevels() == null) return new String[0];
+
+        final BitSet childInclusiveLevels = node.getInclusiveLevels();
+        final int cardinality = childInclusiveLevels.cardinality();
+        List<String> domainLevel = new ArrayList<>(cardinality);
```

Comment:
This method is unfisnished and will be reworked, yet there is no point of having a list here, if the cardinality is known before. An array of ints can be allocated directly. The method should return integers (indices of the levels).